lua/orgmode/ui/virtcookie.lua

```diff
@@ -0,0 +1,258 @@
+  ---@type table<integer, OrgHeadline>
+  local modified_headlines = {}
+  for line = start_line, end_line, 1 do
+    local success, headline = pcall(org.files.get_closest_headline, org.files, { line + 1, 0 })
+    if headline and success then
+      local headline_row, _, _ = headline:node():start()
+      modified_headlines[headline_row] = headline
+    end
+  end
+  for _, headline in pairs(modified_headlines) do
+    self:_set_cookie(headline)
+    local parents = OrgVirtCookie._parent_headlines(headline)
+    for _, parent in ipairs(parents) do
+      self:_set_cookie(parent)
+    end
+  end
+end
+
+---Get whether the watcher is currently attached
+---@return boolean
+function OrgVirtCookie:is_attached()
+  return self.attached
+end
+
+---Get the registered buffer id for the watcher
+---@return integer
+function OrgVirtCookie:get_bufnr()
+  return self.bufnr
+end
+
+---Starts the watcher for the watcher's buffer if it's not already attached
+function OrgVirtCookie:attach()
+  if self.attached then
+    return
+  end
+  self.attached = true
+  vim.b[self.bufnr].org_cookie_mode = true
+
+  watchers[self.bufnr] = self
+  self:_update_cookies_in_range(0, vim.api.nvim_buf_line_count(self.bufnr) - 1)
+
+  -- We cant use  `nvim_set_decoration_provider()` here because of a
+  -- `org.files.get_closest_headline` call made later "modifies" (reloads) some content which
+  -- causes the decoration provider to explode.
+  --
+  -- There's a way around this by playing entirely with the TS Nodes, but this seems fast enough
+  -- so it'll do.
+  vim.api.nvim_buf_attach(self.bufnr, false, {
+    on_lines = function(_, _, _, start_line, _, end_line)
+      if not self.attached then
+        return true
+      end
+
+      vim.schedule(function()
+        if start_line > 0 then
+          -- Sometimes we miss the outer range, so we want to ensure we grab that in those
+          -- scenarios
+          start_line = start_line - 1
+        end
+        self:_update_cookies_in_range(start_line, end_line)
+      end)
+    end,
+    on_reload = function()
+      self:_update_cookies_in_range(0, vim.api.nvim_buf_line_count(self.bufnr) - 1)
+    end,
+    on_detach = function()
+      self:delete()
+    end,
+  })
+end
+
+---Detaches the watcher if it's attached
+function OrgVirtCookie:detach()
+  if not self.attached then
+    return
+  end
+  self.attached = false
+  vim.b[self.bufnr].org_cookie_mode = false
+  vim.api.nvim_buf_clear_namespace(self.bufnr, self.ns_id, 0, -1)
+end
+
+---Toggles the attached state for the watcher
+function OrgVirtCookie:toggle()
+  if self.attached then
+    self:detach()
+  else
+    self:attach()
+  end
+end
+
+---Completely removes the watcher, including from tracked watchers
+function OrgVirtCookie:delete()
+  self:detach()
+  watchers[self.bufnr] = nil
+end
+
+-- We only want to create the autocmds for registering watchers once
+local ft_autocmd_created = false
```

Comment:
I don't see this being used anywhere. And if we need to ensure there's only one autocmd, we can add a clearing au group.